File: apps/android/app/src/main/java/inc/anky/android/core/protocol/AnkyWriter.kt
package inc.anky.android.core.protocol

class AnkyWriter private constructor(
    private val lines: MutableList<String>,
    private var lastAcceptedEpochMs: Long?,
    var isClosed: Boolean,
    var writingElapsedMs: Long,
) {
    constructor() : this(mutableListOf(), null, false, 0)

    val text: String
        get() = lines.joinToString(separator = "\n")

    val isStarted: Boolean
        get() = lines.isNotEmpty()

    val lastAcceptedMs: Long?
        get() = lastAcceptedEpochMs

    fun accept(glyph: String, epochMs: Long): Boolean {
        if (isClosed || !glyph.isSingleProtocolGlyph()) return false
        val protocolGlyph = glyph.protocolGlyphText()
        val deltaMs = lastAcceptedEpochMs?.let { maxOf(0, epochMs - it) }
        val line = if (lastAcceptedEpochMs == null) {
            "$epochMs $protocolGlyph"
        } else {
            "$deltaMs $protocolGlyph"
        }
        lines += line
        writingElapsedMs += deltaMs ?: 0
        lastAcceptedEpochMs = epochMs
        return true
    }

    /**
     * The protocol cannot represent deletions, so a backspace is recorded as a
     * suffix rewrite: keep the first [keepingPrefixGlyphCount] glyphs and re-type
     * the replacement on top. Mirrors iOS AnkyWriter.replaceSuffix — the
     * replacement is never empty, so the text can never go empty through here.
     * Returns the glyphs that were re-typed (empty when nothing changed).
     */
    fun replaceSuffix(keepingPrefixGlyphCount: Int, replacementText: String, epochMs: Long): List<String> {
        if (isClosed) return emptyList()

        val replacementGlyphs = replacementText
            .filterNot { it == '\n' || it == '\r' }
            .protocolGlyphsOrNull()
            .orEmpty()
        if (replacementGlyphs.isEmpty()) return emptyList()

        val currentEvents = absoluteEvents()
        val keptCount = keepingPrefixGlyphCount.coerceIn(0, currentEvents.size)
        val events = currentEvents.take(keptCount).toMutableList()
        val times = replacementEventTimes(
            glyphCount = replacementGlyphs.size,
            anchorEpochMs = events.lastOrNull()?.epochMs,
            originalStartEpochMs = currentEvents.firstOrNull()?.epochMs,
            insertionEpochMs = epochMs,
        )

        replacementGlyphs.forEachIndexed { index, glyph ->
            events += AbsoluteWritingEvent(epochMs = times[index], glyph = glyph)
        }

        applyAbsoluteEvents(events)
        return replacementGlyphs
    }

    fun closeWithTerminalSilence() {
        if (!isStarted || isClosed) return
        lines += AnkyDuration.TerminalSilenceMs.toString()
        isClosed = true
    }

    private data class AbsoluteWritingEvent(
        val epochMs: Long,
        val glyph: String,
    )

    private fun absoluteEvents(): List<AbsoluteWritingEvent> {
        val parsed = runCatching { AnkyParser.parse(text) }.getOrNull() ?: return emptyList()
        var cursor = parsed.startEpochMs
        return parsed.events.mapIndexed { index, event ->
            if (index > 0) cursor += event.deltaMs
            AbsoluteWritingEvent(epochMs = cursor, glyph = event.glyph)
        }
    }

    private fun replacementEventTimes(
        glyphCount: Int,
        anchorEpochMs: Long?,
        originalStartEpochMs: Long?,
        insertionEpochMs: Long,
    ): List<Long> {
        if (glyphCount <= 0) return emptyList()

        if (anchorEpochMs != null) {
            return distributedTimes(
                startEpochMs = anchorEpochMs,
                endEpochMs = maxOf(anchorEpochMs, insertionEpochMs),
                gapCount = glyphCount,
                includesAnchor = false,
            )
        }

        val startEpochMs = minOf(originalStartEpochMs ?: insertionEpochMs, insertionEpochMs)
        if (glyphCount == 1) return listOf(startEpochMs)

        return distributedTimes(
            startEpochMs = startEpochMs,
            endEpochMs = maxOf(startEpochMs, insertionEpochMs),
            gapCount = glyphCount - 1,
            includesAnchor = true,
        )
    }

    private fun distributedTimes(
        startEpochMs: Long,
        endEpochMs: Long,
        gapCount: Int,
        includesAnchor: Boolean,
    ): List<Long> {
        if (gapCount <= 0) {
            return if (includesAnchor) listOf(startEpochMs) else emptyList()
        }

        val elapsed = maxOf(0, endEpochMs - startEpochMs)
        val baseDelta = elapsed / gapCount
        val remainder = (elapsed % gapCount).toInt()
        var cursor = startEpochMs
        val times = if (includesAnchor) mutableListOf(startEpochMs) else mutableListOf()

        for (index in 0 until gapCount) {
            cursor += baseDelta + if (index < remainder) 1 else 0
            times += cursor
        }

        return times
    }

    private fun applyAbsoluteEvents(events: List<AbsoluteWritingEvent>) {
        val first = events.firstOrNull()
        if (first == null) {
            lines.clear()
            writingElapsedMs = 0
            lastAcceptedEpochMs = null
            return
        }

        val rebuiltLines = mutableListOf("${first.epochMs} ${first.glyph.protocolGlyphText()}")
        var previousEpochMs = first.epochMs
        var elapsed = 0L

        for (event in events.drop(1)) {
            val delta = maxOf(0, event.epochMs - previousEpochMs)
            rebuiltLines += "$delta ${event.glyph.protocolGlyphText()}"
            elapsed += delta
            previousEpochMs = event.epochMs
        }

        lines.clear()
        lines += rebuiltLines
        writingElapsedMs = elapsed
        lastAcceptedEpochMs = previousEpochMs
    }

    fun prepareToResume(epochMs: Long) {
        if (!isStarted || isClosed) return
        lastAcceptedEpochMs = epochMs
    }

    companion object {
        fun fromDraft(draftText: String): AnkyWriter {
            val parsed = AnkyParser.parse(draftText)
            val lines = draftText
                .replace("\r\n", "\n")
                .split("\n")
                .dropLastWhile { it.isEmpty() }
                .toMutableList()
            val elapsedWithoutTerminal = parsed.events.sumOf { it.deltaMs }
            return AnkyWriter(
                lines = lines,
                lastAcceptedEpochMs = parsed.startEpochMs + elapsedWithoutTerminal,
                isClosed = parsed.terminalSilenceMs == AnkyDuration.TerminalSilenceMs,
                writingElapsedMs = elapsedWithoutTerminal,
            )
        }
    }
}

private fun String.protocolGlyphText(): String =
    if (this == " ") "SPACE" else this

File: apps/ios/Anky/Core/Protocol/AnkyWriter.swift
import Foundation

public struct AnkyWriter {
    private var lines: [String]
    private var lastAcceptedEpochMs: Int64?
    private(set) public var writingElapsedMs: Int64
    private(set) public var isClosed: Bool

    public init() {
        self.lines = []
        self.lastAcceptedEpochMs = nil
        self.writingElapsedMs = 0
        self.isClosed = false
    }

    public init(draftText: String) throws {
        let parsed = try AnkyParser.parse(draftText)
        self.lines = AnkyWriter.lines(from: draftText)
        self.isClosed = parsed.terminalSilenceMs != nil

        let elapsedWithoutTerminal = parsed.events.reduce(Int64(0)) { sum, event in
            sum + event.deltaMs
        }
        self.writingElapsedMs = elapsedWithoutTerminal
        self.lastAcceptedEpochMs = parsed.startEpochMs + elapsedWithoutTerminal
    }

    public var text: String {
        lines.joined(separator: "\n")
    }

    public var isStarted: Bool {
        !lines.isEmpty
    }

    public var lastAcceptedMs: Int64? {
        lastAcceptedEpochMs
    }

    public mutating func prepareToResume(at epochMs: Int64) {
        guard isStarted, !isClosed else {
            return
        }
        lastAcceptedEpochMs = epochMs
    }

    @discardableResult
    public mutating func accept(_ character: Character, at epochMs: Int64) -> Bool {
        guard !isClosed, isProtocolCharacter(character) else {
            return false
        }

        if let lastAcceptedEpochMs {
            let delta = max(0, epochMs - lastAcceptedEpochMs)
            lines.append("\(delta) \(Self.protocolCharacterText(for: character))")
            writingElapsedMs += delta
        } else {
            lines.append("\(epochMs) \(Self.protocolCharacterText(for: character))")
        }

        lastAcceptedEpochMs = epochMs
        return true
    }

    @discardableResult
    public mutating func replaceSuffix(
        keepingPrefixCharacterCount prefixCount: Int,
        with replacementText: String,
        at epochMs: Int64
    ) -> [Character] {
        guard !isClosed else {
            return []
        }

        let replacementCharacters = Array(replacementText.filter(isProtocolCharacter))
        guard !replacementCharacters.isEmpty else {
            return []
        }

        let currentEvents = absoluteEvents()
        let keptCount = min(max(0, prefixCount), currentEvents.count)
        var events = Array(currentEvents.prefix(keptCount))
        let currentLastEpochMs = currentEvents.last?.epochMs
        let detachedWallCursorMs = detachedWallCursor(
            currentLastEpochMs: currentLastEpochMs,
            insertionEpochMs: epochMs
        )
        let effectiveInsertionEpochMs = detachedWallCursorMs.map { cursor in
            (currentLastEpochMs ?? epochMs) + max(0, epochMs - cursor)
        } ?? epochMs
        let times = replacementEventTimes(
            characterCount: replacementCharacters.count,
            anchorEpochMs: events.last?.epochMs,
            originalStartEpochMs: currentEvents.first?.epochMs,
            insertionEpochMs: effectiveInsertionEpochMs
        )

        for (index, character) in replacementCharacters.enumerated() {
            events.append(AbsoluteWritingEvent(epochMs: times[index], character: character))
        }

        applyAbsoluteEvents(events)
        if detachedWallCursorMs != nil {
            lastAcceptedEpochMs = epochMs
        }
        return replacementCharacters
    }

    public mutating func closeWithTerminalSilence(after terminalSilenceMs: Int64 = AnkyDuration.defaultTerminalSilenceMs) {
        guard isStarted, !isClosed else {
            return
        }
        lines.append("\(AnkyDuration.clampedTerminalSilenceMs(terminalSilenceMs))")
        isClosed = true
    }

    private func isProtocolCharacter(_ character: Character) -> Bool {
        character != "\n" && character != "\r"
    }

    private static func protocolCharacterText(for character: Character) -> String {
        character == " " ? "SPACE" : String(character)
    }

    private struct AbsoluteWritingEvent {
        let epochMs: Int64
        let character: Character
    }

    private func absoluteEvents() -> [AbsoluteWritingEvent] {
        guard let parsed = try? AnkyParser.parse(text) else {
            return []
        }

        var cursor = parsed.startEpochMs
        return parsed.events.enumerated().map { index, event in
            if index > 0 {
                cursor += event.deltaMs
            }
            return AbsoluteWritingEvent(epochMs: cursor, character: event.character)
        }
    }

    private func replacementEventTimes(
        characterCount: Int,
        anchorEpochMs: Int64?,
        originalStartEpochMs: Int64?,
        insertionEpochMs: Int64
    ) -> [Int64] {
        guard characterCount > 0 else {
            return []
        }

        if let anchorEpochMs {
            return distributedTimes(
                from: anchorEpochMs,
                to: max(anchorEpochMs, insertionEpochMs),
                gapCount: characterCount,
                includesAnchor: false
            )
        }

        let startEpochMs = min(originalStartEpochMs ?? insertionEpochMs, insertionEpochMs)
        guard characterCount > 1 else {
            return [startEpochMs]
        }

        return distributedTimes(
            from: startEpochMs,
            to: max(startEpochMs, insertionEpochMs),
            gapCount: characterCount - 1,
            includesAnchor: true
        )
    }

    private func detachedWallCursor(currentLastEpochMs: Int64?, insertionEpochMs: Int64) -> Int64? {
        guard let lastAcceptedEpochMs,
              let currentLastEpochMs,
              lastAcceptedEpochMs > currentLastEpochMs,
              insertionEpochMs >= lastAcceptedEpochMs else {
            return nil
        }
        return lastAcceptedEpochMs
    }

    private func distributedTimes(
        from startEpochMs: Int64,
        to endEpochMs: Int64,
        gapCount: Int,
        includesAnchor: Bool
    ) -> [Int64] {
        guard gapCount > 0 else {
            return includesAnchor ? [startEpochMs] : []
        }

        let elapsed = max(0, endEpochMs - startEpochMs)
        let baseDelta = elapsed / Int64(gapCount)
        let remainder = Int(elapsed % Int64(gapCount))
        var cursor = startEpochMs
        var times = includesAnchor ? [startEpochMs] : []

        for index in 0..<gapCount {
            cursor += baseDelta + (index < remainder ? 1 : 0)
            times.append(cursor)
        }

        return times
    }

    private mutating func applyAbsoluteEvents(_ events: [AbsoluteWritingEvent]) {
        guard let first = events.first else {
            lines = []
            writingElapsedMs = 0
            lastAcceptedEpochMs = nil
            return
        }

        var rebuiltLines = ["\(first.epochMs) \(Self.protocolCharacterText(for: first.character))"]
        var previousEpochMs = first.epochMs
        var elapsed: Int64 = 0

        for event in events.dropFirst() {
            let delta = max(0, event.epochMs - previousEpochMs)
            rebuiltLines.append("\(delta) \(Self.protocolCharacterText(for: event.character))")
            elapsed += delta
            previousEpochMs = event.epochMs
        }

        lines = rebuiltLines
        writingElapsedMs = elapsed
        lastAcceptedEpochMs = previousEpochMs
    }

    private static func lines(from text: String) -> [String] {
        var lines = text
            .replacingOccurrences(of: "\r\n", with: "\n")
            .split(separator: "\n", omittingEmptySubsequences: false)
            .map(String.init)

        if lines.last == "" {
            lines.removeLast()
        }

        return lines
    }
}
